src/textual/widgets/_tabbed_content.py

```diff
@@ -0,0 +1,148 @@
+
+    This widget is intended to be used with [TabbedContent][textual.widgets.TabbedContent].
+
+    """
+
+    DEFAULT_CSS = """
+    TabPane {
+        height: auto;
+        padding: 1 2;
+    }
+    """
+
+    def __init__(
+        self,
+        title: TextType,
+        *children: Widget,
+        name: str | None = None,
+        id: str | None = None,
+        classes: str | None = None,
+        disabled: bool = False,
+    ):
+        """Initialize a TabPane.
+
+        Args:
+            title: Title of the TabPane (will be displayed in a tab label).
+            name: Optional name for the TabPane.
+            id: Optional ID for the TabPane.
+            classes: Optional initial classes for the widget.
+            disabled: Whether the TabPane is disabled or not.
+        """
+        self._title = self.render_str(title)
+        super().__init__(
+            *children, name=name, id=id, classes=classes, disabled=disabled
+        )
+
+
+class TabbedContent(Widget):
+    """A container with associated tabs to toggle content visibility."""
+
+    DEFAULT_CSS = """
+    TabbedContent {
+        height: auto;
+    }
+    TabbedContent > ContentSwitcher {
+        height: auto;
+    }
+    """
+
+    active: reactive[str] = reactive("", init=False)
+    """The ID of the active tab, or empty string if none are active."""
+
+    def __init__(self, *titles: TextType, initial: str = "") -> None:
+        """Initialize a TabbedContent widgets.
+
+        Args:
+            *titles: Positional argument will be used as title.
+            initial: The id of the initial tab, or empty string to select the first tab.
+        """
+        self.titles = [self.render_str(title) for title in titles]
+        self._tab_content: list[Widget] = []
+        self._initial = initial
+        super().__init__()
+
+    def compose(self) -> ComposeResult:
+        """Compose the tabbed content."""
+
+        def set_id(content: TabPane, new_id: str) -> TabPane:
+            """Set an id on the content, if not already present."""
+            if content.id is None:
+                content.id = new_id
+            return content
+
+        # Wrap content in a `TabPane` if required.
+        pane_content = [
+            (
+                set_id(content, f"tab-{index}")
+                if isinstance(content, TabPane)
+                else TabPane(
+                    title or self.render_str(f"Tab {index}"), content, id=f"tab-{index}"
+                )
+            )
+            for index, (title, content) in enumerate(
+                zip_longest(self.titles, self._tab_content), 1
+            )
+        ]
+        # Get a tab for each pane
+        tabs = [
+            ContentTab(content._title, content.id or "") for content in pane_content
+        ]
+        # Yield the tabs
+        yield Tabs(*tabs, active=self._initial)
+        # Yield the content switcher and panes
+        with ContentSwitcher(initial=self._initial):
```

Comment:
I think an extra step is required here, where `self._initial` needs to be tested if it's empty and, if so, it's set to the `id` of the first tab.

Reply:
The first tab is set via the TabActivated event.